### PR TITLE
add cse.d: manage CSE memory in stack frame

### PR DESCRIPTION
**Pull request by WalterBright**
The distributed nature of https://github.com/dlang/dmd/pull/9905 suggests that managing CSEs can be encapsulated into its own module. This is a start.

### CODE REVIEW

File: src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -2294,6 +2295,7 @@ regm_t cod3_useBP()
            )
             goto Lcant;
     }
+Lcan:
```
Comment (thewilsonator) [inline src/dmd/backend/cod3.d]: ?

**Comment by thewilsonator** [comment]
Please add `nothrow` throughout.

File: src/dmd/backend/cgcse.d [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * Copyright:   Copyright (C) 1985-1998 by Symantec
+ *              Copyright (C) 2000-2019 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/cgcse.d, backend/cgcse.d)
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/backend/cgcse.d
+ */
+
+module dmd.backend.cgcse;
```
Comment (jacob-carlborg) [inline src/dmd/backend/cgcse.d]: I suggest you take the opportunity to give this new module a better less cryptic name for those not familiar with the code base. 

BTW, all these `cg` prefixes are a good indication that a package would be better.

Comment (WalterBright) [reply src/dmd/backend/cgcse.d]: CSE is commonplace compiler jargon. If you still need to know what it is, read the header comment in the file.

Comment (thewilsonator) [reply src/dmd/backend/cgcse.d]: True, but `cgcse` looks like someone had a seizure at the keyboard. Also CSE is an optimisation pass and has nothing to do with codegen, which is what I associate with `cg`.

File: src/dmd/backend/cgcse.d [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+/****************************
+ * Table of common subexpressions stored on the stack.
+ *      csextab[]       array of info on saved CSEs
+ *      CSEpe           pointer to saved elem
+ *      CSEregm         mask of register that was saved (so for multi-
+ *                      register variables we know which part we have)
+ */
+
+enum CSEload       = 1;       // set if the CSE was ever loaded
```
Comment (Basile-z) [inline src/dmd/backend/cgcse.d]: Public declaration so DDOC style slashes comments. This comment counts also for pretty much everything that's commented in the module (unless you meant to set `private`  CSE fields)

File: src/dmd/backend/cgcse.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/backend/cgcse.d]: > CSE is an optimisation pass and has nothing to do with codegen

As you can see the codegen is very much involved with CSE.

File: src/dmd/backend/cgcse.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/cgcse.d]: For easy review, I kept this code movement as cut/paste. It's really hard to review such if it is all refactored as well as being moved. As I explain again and again, small incremental PRs are far easier to review than large ones, and far easier to debug if a regression is introduced. In light of that, this should be reviewed as progress towards a goal, not whether the goal is achieved.

File: src/dmd/backend/cgcse.d (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/backend/cgcse.d]: The code doesn't look like it does much at all. 
CSE is usually an IR to IR pass, I don't know the backend is doing if its not that.

Comment (WalterBright) [reply src/dmd/backend/cgcse.d]: > The code doesn't look like it does much at all.

That's right, for a first step.

_For easy review, I kept this code movement as cut/paste. It's really hard to review such if it is all refactored as well as being moved. As I explain again and again, small incremental PRs are far easier to review than large ones, and far easier to debug if a regression is introduced. In light of that, this should be reviewed as progress towards a goal, not whether the goal is achieved._

> CSE is usually an IR to IR pass

Part of it is, see gdag.d The rest is in the code generator.

**Comment by thewilsonator** [comment]
```
dmd\backend\cgcod.d(2324): Error: function `dmd.backend.cgcse.CSE.updateSizeAndAlign` is not `nothrow`
dmd\backend\cgcod.d(2265): Error: `nothrow` function `dmd.backend.cgcod.cse_save` may throw
```

File: src/dmd/backend/cgcse.d [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  __gshared:
+    uint slotSize;          // size of each slot in table
+    uint alignment;         // alignment for the table
+
+    /********************************
+     * Update slot size and alignment to worst case.
+     * A bit wasteful of stack space.
+     */
+    static void updateSizeAndAlign(elem* e)
```
Comment (thewilsonator) [inline src/dmd/backend/cgcse.d]: needs to be `nothrow`.

**Review by thewilsonator — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
@WalterBright please follow up with a PR that updates the readme: https://github.com/dlang/dmd/blob/master/src/dmd/backend/README.md

**Pull request merged**